base/strings/annotated.jl

```diff
@@ -386,3 +386,132 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
 Get all annotations of `chr`.
 """
 annotations(c::AnnotatedChar) = c.annotations
+
+## AnnotatedIOBuffer
+
+struct AnnotatedIOBuffer <: AbstractPipe
+    io::IOBuffer
+    annotations::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}
+end
+
+AnnotatedIOBuffer(io::IOBuffer) = AnnotatedIOBuffer(io, Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}())
+AnnotatedIOBuffer() = AnnotatedIOBuffer(IOBuffer())
+
+function show(io::IO, aio::AnnotatedIOBuffer)
+    show(io, AnnotatedIOBuffer)
+    print(io, '(', aio.io.size, " byte", ifelse(aio.io.size == 1, "", "s"), ", ",
+          length(aio.annotations), " annotation", ifelse(length(aio.annotations) == 1, "", "s"), ")")
+end
+
+pipe_reader(io::AnnotatedIOBuffer) = io.io
+pipe_writer(io::AnnotatedIOBuffer) = io.io
+
+# Useful `IOBuffer` methods that we don't get from `AbstractPipe`
+position(io::AnnotatedIOBuffer) = position(io.io)
+seek(io::AnnotatedIOBuffer, n::Integer) = (seek(io.io, n); io)
+seekend(io::AnnotatedIOBuffer) = seekend(io.io)
+skip(io::AnnotatedIOBuffer, n::Integer) = (skip(io.io, n); io)
+copy(io::AnnotatedIOBuffer) = AnnotatedIOBuffer(copy(io.io), copy(io.annotations))
+
+annotations(io::AnnotatedIOBuffer) = io.annotations
+
+function write(io::AnnotatedIOBuffer, astr::Union{AnnotatedString, SubString{<:AnnotatedString}})
+    astr = AnnotatedString(astr)
+    offset = position(io.io)
+    if !eof(io)
+        # If we are overwriting an existing span in the AnnotatedIOBuffer,
+        # clear out any overlapping pre-existing annotations.
+        span = offset+1:offset+ncodeunits(astr)
+        filter!(((range, _),) -> first(range) < first(span) || last(range) > last(span), io.annotations)
+        extras = Tuple{UnitRange{Int}, Pair{Symbol, Any}}[]
+        for i in eachindex(io.annotations)
+            range, annot = io.annotations[i]
+            # Test for partial overlap
+            if first(range) <= first(span) <= last(range) || first(range) <= last(span) <= last(range)
+                io.annotations[i] = (if first(range) < first(span)
+                                         first(range):first(span)-1
+                                     else last(span)+1:last(range) end, annot)
+                # If `span` fits exactly within `range`, then we've only copied over
+                # the beginning overhang, but also need to conserve the end overhang.
+                if first(range) < first(span) && last(span) < last(range)
+                    push!(extras, (last(span)+1:last(range), annot))
+                end
+            end
+            # Insert any extra entries in the appropriate position
+            for entry in extras
+                indices = searchsorted(io.annotations, (first(entry),), by=first)
+                splice!(io.annotations, indices, Tuple{UnitRange{Int}, Pair{Symbol, Any}}[entry])
```

Comment:
Do you mean to try to overwrite existing annotations here, or do you want to just insert the element at that place (which may not require as many temporary lists to be allocated):
```suggestion
                atwhere = searchsortedlast(io.annotations, (first(entry),), by=first)
                insert!(io.annotations, atwhere, entry)
```

Reply:
I indeed just want to insert annotations in the sorted location. Thanks!

Reply:
Ah, the `splice!` -> `insert!` change may not be viable after all (or needs a `max(1, _)`). Looking at the tests:

```
Error During Test at /home/tec/Documents/Projects/julia/test/strings/annotated.jl:3
  Got exception outside of a @test
  BoundsError: attempt to access 0-element Vector{Tuple{UnitRange{Int64}, Pair{Symbol, Any}}} at index [0]
  Stacktrace:
    [1] _growat!(a::Vector{Tuple{UnitRange{Int64}, Pair{Symbol, Any}}}, i::Int64, delta::Int64)
      @ Base ./array.jl:1128
    [2] insert!
      @ ./array.jl:1727 [inlined]
    [3] annotate!(s::Base.AnnotatedString{String}, range::UnitRange{Int64}, labelval::Pair{Symbol})
      @ Base ./strings/annotated.jl:334
```

Reply:
Ah, seems like a `+ 1` is needed. Then the tests are happy, and so I've just pushed this version.